### PR TITLE
doc: external storage of encryption passphrases

### PR DESCRIPTION
**Pull request by tareksha**
# Describe what this PR does #

Proposal for https://github.com/ceph/ceph-csi/issues/2209

Depends-On: #2752

### CODE REVIEW

**Comment by yati1998** [comment]
@tareqhs please follow the proper [commit guidelines](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages) for avoid DCO error.

File: docs/design/proposals/kms-custom-http.md
Comment (nixpanic) [inline docs/design/proposals/kms-custom-http.md:39]: In case you would enable `basicauth`, where/how would a user configure login details?

File: docs/design/proposals/kms-custom-http.md
Comment (nixpanic) [inline docs/design/proposals/kms-custom-http.md]: Can there be a simple open source service provided that ceph-csi can use for automated e2e testing? A container-image that the e2e tests can deploy would be ideal.

File: docs/design/proposals/kms-custom-http.md
Comment (humblec) [inline docs/design/proposals/kms-custom-http.md:25]: may be blackboxkmsservice ? :)

File: docs/design/proposals/kms-custom-http.md
Comment (humblec) [reply docs/design/proposals/kms-custom-http.md:39]: rather curious to know how its going to make a decision/logic  behind `enabled basicauth`?

Comment (nixpanic) [reply docs/design/proposals/kms-custom-http.md:39]: I assume it is https://datatracker.ietf.org/doc/html/rfc7617 and needs options for usersname/password

File: docs/design/proposals/kms-custom-http.md
Comment (nixpanic) [reply docs/design/proposals/kms-custom-http.md:25]: or just `kvstore` as the API seems a very simple key/value store?

File: docs/design/proposals/kms-custom-http.md
Comment (tareksha) [reply docs/design/proposals/kms-custom-http.md:39]: I assume like aws kms, in a separate secret. I would prefer to define the behavior in the config and not guess based on runtime basic auth challenge responses.

File: docs/design/proposals/kms-custom-http.md
Comment (tareksha) [reply docs/design/proposals/kms-custom-http.md]: A good suggestion, but the differentiator here is to completely customize the API and not align on predefined schemes. for predefined APIs for this we already have aws, vault, ...

File: docs/design/proposals/kms-custom-http.md
Comment (tareksha) [reply docs/design/proposals/kms-custom-http.md:25]: i would go with @nixpanic's suggestion here

File: docs/design/proposals/kms-custom-http.md
Comment (Rakshith-R) [reply docs/design/proposals/kms-custom-http.md:39]: >I assume like aws kms, in a separate secret. I would prefer to define the behavior in the config and not guess based on runtime basic auth challenge responses.

@tareqhs Adding an example of such a secret/config in design doc would help explain this better. 

You may also be interested in selecting different such secret/configs according to kubernetes namespace / tenant where PVC was created like in other encryptions currently available (just a suggestion).

**Comment by Rakshith-R** [comment]
Please address the ci `commitlint,lint-extras,DCO` errors.

File: docs/design/proposals/kms-custom-http.md
Comment (nixpanic) [reply docs/design/proposals/kms-custom-http.md]: The KMS implementation uses defined endpoints as mentioned at the bottom of the document. A very simple KMS service would help to validate the implementation and functioning of Ceph-CSI. The KMS service could run completely in-memory, with a simple `map[string]string` kind of structure that handles the endpoints.

**Comment by yati1998** [comment]
@tareqhs please do address the commitlint issue.

**Comment by tareksha** [comment]
@yati1998 , @Rakshith-R , fixed

File: docs/design/proposals/kms-custom-http.md
Comment (tareksha) [reply docs/design/proposals/kms-custom-http.md]: sure, for validation it would be very convenient

**Comment by yati1998** [comment]
@tareqhs please check the lint-extras errors and resolve them. There are few rules that we need to take care of. A detailed description is mentioned in the logs, do let me know if you face any issues.

**Comment by tareksha** [comment]
@yati1998 done

**Comment by yati1998** [comment]
@tareqhs please follow the proper commit guidelines for avoid commitlint error.
the commit message can be something like:
```
doc: external storage of encryption passphrases

description

<signed-off-by>
```
Go through the link mentioned in the comment above.

**Comment by yati1998** [comment]
And also the lint-extra issues are not yet resolved. Go through the details and make changes accordingly.

**Comment by tareksha** [comment]
@yati1998 done

File: docs/design/proposals/kms-custom-http.md
Comment (nixpanic) [reply docs/design/proposals/kms-custom-http.md:25]: yes please!

**Review by nixpanic — CHANGES_REQUESTED** [review]
Sorry for the late responses! Could you address the name of the KMS service and create a single commit (use `git rebase -i ..` and `fixup`), then force-push this PR again?

Thanks!

**Comment by nixpanic** [comment]
@mergifyio refresh

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio refresh

**Comment by humblec** [comment]
@Mergifyio refresh

**Comment by humblec** [comment]
added DNM and refresh to move the PR from the merge queue which does not looks like helping.. Force push or close and reopen may help .. any other way to move the PR from the merge queue?  @ceph/ceph-csi-contributors

**Comment by humblec** [comment]
@tareqhs I am closing and reopening this PR  ( wihtout your permission :) ) just to allow other PRs in the merge queue to proceed as this is blocking other PRs in the merge queue due to mergify rules changes we did today.. Apologies in advance.

**Comment by nixpanic** [comment]
@mergifyio refresh

**Comment by nixpanic** [comment]
@mergifyio rebase

It seems this PR is still blocking the merge queue (not approved yet, but somehow in the queue until approved/merged?).

**Comment by humblec** [comment]
> @Mergifyio rebase
> 
> It seems this PR is still blocking the merge queue (not approved yet, but somehow in the queue until approved/merged?).

Looks like none of the actions here helps to move this out of merge queue... @nixpanic , it is still in the queue as #1

**Pull request closed without merging**
